### PR TITLE
New Reply/Canned Responses library for replying to messages

### PR DESCRIPTION
**Pull request by WhelanB**
Adding a new library, "Reply", which allows apps to prompt the user to either select from a set of canned responses or to compose a reply if a ```textinput``` keyboard app is installed. The library handles constructing the reply object for Gadgetbridge and can optionally send this via ```Bluetooth.println``` or handle the reply directly, if they wish to use it for other purposes

It can be used by calling:

```require("reply").reply(options).then(...```

from an app.

Canned responses can be set from the App Loader - though using emojis seems to break the created JSON and I haven't looked at re-encoding the text first before writing yet.

Untitled
image

After it merges, I might go ahead and add it to the default messages app - I've been testing it personally for the past week or so and can reply to messages directly from my Bangle, which is a nice feature to have!

### CODE REVIEW

File: apps/reply/interface.html [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+        }
+
+        function toggleVisibility() {
+            let empty = document.getElementById("empty");
+            let loading = document.getElementById("loading");
+            if (replies.length == 0) {
+                empty.setAttribute("class", "d-block");
+                loading.setAttribute("class", "d-hide");
+            }
+            else if (fetching) {
```
Comment (bobrippling) [inline apps/reply/interface.html:68]: Should these be the other way round? Testing `fetching`, then `replies.length` ?

File: apps/reply/interface.html [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+        function updateDevice() {
+            fetching = true;
+            renderReplyList();
+            Util.writeStorage("replies.json", JSON.stringify(replies), () => {
+                fetching = false;
+                renderReplyList();
+                return true;
+            });
+            fetching = false;
```
Comment (bobrippling) [inline apps/reply/interface.html:119]: Should the `fetching = false` above (line 115) be the one to set `fetching` to false, and not this line?

File: apps/reply/lib.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  }
+
+  return new Promise((resolve, reject) => {
+    var menu = {
+      "": {
+        title: options.title || /*LANG*/ "Reply with:",
+        back: function () {
+          E.showMenu();
+          layout.setUI();
+          layout.render();
```
Comment (bobrippling) [inline apps/reply/lib.js:30]: Do we need to do these when the user's exiting?

File: apps/reply/lib.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+    if (replies.length == 0) {
+      if (!keyboard) {
+        E.showPrompt(
+          /*LANG*/ "Please install a keyboard app, or set a custom reply via the app loader!",
+          {
+            buttons: { Ok: true },
+            remove: function () {
+              layout.setUI();
+              layout.render();
```
Comment (bobrippling) [inline apps/reply/lib.js:60]: Same here :)

**Review by bobrippling — COMMENTED** [review]
Looks great! Just a few comments, looks like the linters picked up on `layout` not being declared too - let me know if you want a hand with any of this

File: apps/reply/lib.js (hunk 3 above)
Comment (WhelanB) [reply apps/reply/lib.js]: I had added this to fix a white screen during testing - just removed it as the calling app can redraw after the promise resolves!

File: apps/reply/interface.html (hunk 2 above)
Comment (WhelanB) [reply apps/reply/interface.html]: Yes you're right, fetching should only update when the promise resolves - fixed!

File: apps/reply/lib.js (hunk 3 above)
Comment (bobrippling) [reply apps/reply/lib.js]: Cool, sounds good!

**Comment by bobrippling** [comment]
Thanks, looking forward to seeing it see some action!

**Comment by soneret** [comment]
great job 
 i cant see reply function when massage arrived .i installed your app.what can be problem .thanks in advance

**Comment by WhelanB** [comment]
> great work thanks i cant see reply function when massage arrived .i installed your app.what can be problem .thanks in advance

I still need to open a second PR to use the library from inside the Messages app - I've been testing it myself but haven't had a chance to publish it yet!

**Pull request merged**